Minimal changes to get building on OS X with clang and Homebrew

taoeffect opened the pull request:
Install dependencies with Homebrew and then from within src directory:

```
make USE_UPNP=1 -j $(sysctl -n hw.ncpu) -f Makefile.osx all
```

The only concern is how to install specific versions of the dependencies (to avoid accidentally resulting a hard fork).

Unfortunately, I couldn't find "db-4.7.25.NC" via Homebrew. The farthest back they have is 4.8.30. Is that OK?

Also, I haven't tested yet if boost 1.5 builds on OS X or not. 1.55 does, but I'm not sure about 1.5.

taoeffect committed:
This should build on OS X. Use homebrew to install deps & this command:

Within src directory:

    make USE_UPNP=1 -j $(sysctl -n hw.ncpu) -f Makefile.osx all

taoeffect commented:
Note also that this is just for building `namecoind` (not the QT GUI).

Midar commented:
This would be obsoleted if https://github.com/namecoin/namecoin/pull/135 is merged

taoeffect commented:
:+1: @Midar. [commented](https://github.com/namecoin/namecoin/pull/135#issuecomment-48990044)

taoeffect commented:
Update: this was fixed. Safe to merge!

~~Also, just an FYI, this does contain a breaking change for some compilers. If someone needs me to fix that just let me know and I'll figure out the macro for detecting clang compilers (instead of just commenting [this section](https://github.com/namecoin/namecoin/pull/136/files#diff-1fc2d3d7edc00ab8ea29eb1ca30cdcbbR942) out).~~

Midar commented:
PR #108 doesn't break compilation with non-Clang compilers BTW.

cassiniNMC commented:
```
LIBS+= \
 -Wl,-dynamic \ 
 -lgthread-2.0 \    
```

Do we really need gthread?

Midar commented:
The other Makefiles have that, too, but it shouldn't be necessary. My PR that introduces a proper build system doesn't have it and it works :).

indolering commented:
So this does NOT work for Namecoin-QT, correct?

taoeffect commented:
@indolering Correct.

taoeffect commented:
I don't expect this to be merged btw. It's just an illustration of the minimal changes needed to build on OS X.

indolering commented:
@taoeffect Yeah, I had it working ~8 months ago, then changes to 10.9 and the switch to Clang broke Boost.

phelixbtc commented:
It seems OS X is even worse than windows in regard to building. Somebody please make it work! With GUI!
@taoeffect Windows version is also using db-4.8.30.NC

Midar commented:
@phelixbtc I got it working _with_ GUI, see my other PR.

indolering commented:
@Midar https://github.com/ceptacle/libcoinqt Interested?

indolering commented:
@phelixbtc could we get this PR request accepted?  Namecoind is better than nothing and Midar's other build system isn't going anywhere.

taoeffect commented:
What's libcoind missing that namecoind has? I like libcoind's build system better (and I got it to build on OS X, I should probably submit a PR for that too...)

JeremyRand commented:
@taoeffect libcoind is missing a few things, including RPC calls for registering names.  There's more info on the forum.  I might be able to arrange some bounties for getting libcoind up to par with namecoind feature-wise, if that's something you're interested in.

taoeffect commented:
I wouldn't be able to do it (allergic to C++), but I might be able to chip in. Grumble. What about bitcore? Can it do namecoiny things yet? I think @ryancdotorg was working on that...?

taoeffect commented:
NM, @ryancdotorg said on IRC that he's not entirely sure what bitcore is... lol. And I can't figure it out either! Is it a liteclient? It doesn't seem to be a full node.

taoeffect commented:
OK, so figured out what bitcore is thanks to `jgarzik` on `#bitcore`.

Bitcore is:
- A library for doing bitcoin things
- It is _not_ a client.
- The client for bitcore seems to be [copay](https://github.com/bitpay/copay).

Copay is:
- A "wallet".
- _Not_ a full node, or any kind of a node for that matter.
- It uses [insight-api](https://github.com/bitpay/insight-api)

Insight-API seems to be:
- JSON-RPC client to `bitcoind`
- Plus some type of caching database on top of that (I'm guessing).

So... nm about the bitcore. Looks like all we have is `namecoind` and `libcoin --namecoin`, so it seems we'd better focus on `libcoind`.

domob1812 commented:
Sorry for the OT, but I think I should comment here regarding libcoin:  Yes, exactly, the missing features are registering names and `name_list`.  I'm actually planning to start working on these "really soon" (next week at the latest), so hopefully they will be implemented in the not too distant future.  Bounties would, of course, be appreciated, but I'm determined to do it nevertheless.

phelixbtc commented:
Is there a consensus this is a step in the right direction? @cassiniNMC?

@taoeffect Could you give example lines for how to install dependencies with proper versions with homebrew?  Why did you not expect this to be merged?

taoeffect commented:
> @taoeffect Could you give example lines for how to install dependencies with proper versions with homebrew?

Like I said, the "proper versions" aren't available on Homebrew, so I used what was available there. `boost` 1.55 seems to be working. I also installed openssl via Homebrew, as well as `berkeley-db`. These were installed just using regular `brew install [blah]` commands (I think). Try this out and let me know if you run into problems (I'm unfortunately rather pressed for time).

> Why did you not expect this to be merged?

Because [this comment](https://github.com/namecoin/namecoin/pull/136#issuecomment-49118171).

phelixbtc commented:
@taoeffect OK, we should be able figure these out, thanks!

cassiniNMC commented:
> this does contain a breaking change for some compilers. If someone needs me to fix that just let me know and I'll figure out the macro for detecting clang compilers

@taoeffect Yes, it would be great if you could find out a suitable macro/pragma for that section in serialize.h

indolering commented:
> Bounties would, of course, be appreciated, but I'm determined to do it nevertheless.

Please don't say this, you yourself said that you have 5-10 hours of contract work per week: bounties enable you to work on Namecoin.  It's nice that you squeeze it in anyway, but the impact is greater than what you lead people to think.

taoeffect commented:
@cassiniNMC: if either you or @phelixbtc or anyone else can just tell me what the pragma is i'll paste it in there and push a commit. right now i've a billion things going on that are higher priority, and this should be a simple google search for ya'll (and me, admittedly, it's just that for me the google search must be put off until I knock out the high priority items).

phelixbtc commented:
```
#ifdef __WXMSW__
    // Windows
#else
#ifdef MAC_OSX
    // MAC
#else
    // ...
#endif
#endif
```

taoeffect commented:
Hmm, that assumes `clang` isn't being used anywhere but OS X. Isn't it possible these issues could happen on Linux with `clang`?

indolering commented:
@taoeffect the only reason we are using clang on OS X is because Apple fucking replaced GCC specific commands with Clang aliases and removed GCC from OS X 10.9.  If someone wants to use Clang on Linux, that's their problem.

taoeffect commented:
> @taoeffect the only reason we are using clang on OS X is because Apple fucking replaced GCC specific commands with Clang aliases and removed GCC from OS X 10.9. If someone wants to use Clang on Linux, that's there problem.

mmk... that is because clang is better than GCC. :P

OK, I decided to google it. [It's `__clang__`](http://nadeausoftware.com/articles/2012/10/c_c_tip_how_detect_compiler_name_and_version_using_compiler_predefined_macros).

I'll try to get a commit pushed shortly.

taoeffect commented:
Um, somebody please remind me: this branch is pretty far behind `namecoinq` now, so what should I do... merge those changes on top of this commit, rebase (too late probably), just make my changes in a new commit after this one and push without merging, or close PR and re-open after merging?

cc @phelixbtc @cassiniNMC @JeremyRand @domob1812

taoeffect commented:
I feel like (1) merge, (2) make changes, commit, and push is the right answer, but am not sure.

taoeffect committed:
Merge branch 'namecoinq' into macosx

taoeffect committed:
Macro guards to make it compile on non-clang compilers.

taoeffect commented:
OK, impatient so I went ahead with merge, fix, commit, push.

Should be safe to merge now!

taoeffect commented:
Please test this to make sure it compiles on Linux/Windows before merging! I verified it works on OS X.

indolering commented:
@taoeffect Does this require `berkeley-db4` to be installed or will db5 work?  Is this working with QT?

taoeffect commented:
> @taoeffect Does this require `berkeley-db4` to be installed or will db5 work? Is this working with QT?

I actually compiled this with db5... but it should probably be done with db4, right @JeremyRand @ryancdotorg? (cc @domob1812).

Should I change it to use db4 or leave it as-is for use with db5 (which is in homebrew for `berkeley-db`).

indolering commented:
> I actually compiled this with db5... but it should probably be done with db4, right @JeremyRand @ryancdotorg? (cc @domob1812).
> 
> Should I change it to use db4 or leave it as-is for use with db5 (which is in homebrew for berkeley-db).

@taoeffect, well, it's [not building with db4](https://gist.github.com/indolering/4fb6c098d344b8663dc4).

I'm not sure if I should put the following instructions in docs/build-osx.txt or the main readme.md, so I'm going to put it here:
1. Install dependencies using Brew:
   
   brew install openssl boost miniupnpc berkeley-db
2. Execute make command from the `src` directory:
   
   cd ./src
   make USE_UPNP=1 -j $(sysctl -n hw.ncpu) -f Makefile.osx all
3. Standard compiler warnings will appear but if the build is successful you will find a binary at `./src/namecoind`.

taoeffect commented:
> @taoeffect, well, it's not building with db4.

Yeah I know, I need to change the Makefile for that to work. The paths to the libraries and headers need to be manually specified for db4 (like I've done for openssl, for example).

> brew install openssl boost miniupnpc berkeley-db

Yep, except it would be `berkeley-db4` if the others think that's what it should be.

domob1812 commented:
Actually, I would have preferred to rebase (`git rebase` instead of `git merge`) your commits on top of the current branch, so as to not have a "merge" commit in the history.  But it doesn't matter for now - just remember this for the future. ;)

Apart from that, I'm fine with the changes as long as everyone confirms it works for them and fixes the problems on OS X.

indolering commented:
@taoeffect if moving to db4 is easy, just do it.  It can't hurt.

phelixbtc commented:
IIRC database format has changed between db4 and db5 so it is important to always use db4 for compatibility.

Compiles fine on Windows.

> I'm not sure if I should put the following instructions in docs/build-osx.txt or the main readme.md

build-osx.txt imho but I suggest a separate commit to get this one merged.

taoeffect commented:
OK, I'll push an update (hopefully today if I don't forget) to change the Makefile to use db4.

taoeffect committed:
Switched to Berkeley DB 4. Other changes (see details below!!!)

- got rid of gthread
+ removed a bunch of unused and irrelevant-to-osx stuff from Makefile
+ no need to specify `USE_UPNP=1`, it is done for you (make sure to install it!)

Instructions:

1. `brew install openssl berkeley-db4 boost miniupnpc`
2. cd src && make -f Makefile.osx -j $(sysctl -n hw.ncpu) all

taoeffect commented:
K. This is ready to be merged. I got rid of `gthread` and am requiring miniupnpc to be installed (cause... there's no reason not to), simplified Makefile.osx somewhat, and switched to berkeley-db4.

Compiled and verified.

## Will crash if you have previously used this PR to run namecoind because of database downgrade to berkeley db 4!!!

You have two options:
- Figure out how to downgrade the berkeley-db from 5 to 4. (I couldn't figure it out).

Or
- Export your private keys using this command on all of the addresses you own: `./namecoind dumpprivkey <address>`

If you go with the second option, you will need to delete all your database stuff and re-download. You can use my node on okturtles.com to "quickly" get up to speed (I have bandwidth to spare):

```
./namecoind -gen=0 -server -addnode=192.184.93.146
```

Then after it boots up add re-import your private keys that you exported:

```
./namecoind importprivkey 23rjkjf823jf98jsfj8932jfj98fw982jf982jf9fj8
```

_(Obviously, the key above isn't a real one, just an example.)_

taoeffect commented:
Also, regarding `build-osx.txt`, instructions are simple now:
1. `brew install openssl berkeley-db4 boost miniupnpc`
2. `cd src && make -f Makefile.osx -j $(sysctl -n hw.ncpu) all`

If you know your CPU count, you can just use it: `make -f Makefile.osx -j 8 all`

taoeffect commented:
> Actually, I would have preferred to rebase (git rebase instead of git merge) your commits on top of the current branch, so as to not have a "merge" commit in the history. But it doesn't matter for now - just remember this for the future. ;)

@domob1812 I decided to not do a `rebase` because I had already pushed the changes publicly. You aren't supposed to rebase if you've made your stuff public (as that will cause history conflicts).

domob1812 commented:
@taoeffect Considering that this branch on your repo is just a "work in progress", I see no problem with rebasing it.  This is also what Bitcoin devs usually want, and it makes the official repo's history cleaner.  I think this counts more than following some "supposed to be" Git guidelines to-the-letter.

luke-jr commented on src/util.h:
```diff
@@ -109,12 +109,13 @@ T* alignup(T* p)
     return u.ptr;
 }
 
+#define MSG_NOSIGNAL        0
```
Won't this break Linux builds?

taoeffect replied:
Um, I don't know. Is `__WXMSW__` Windows only? Then maybe there should be a separate ifdef that defines it for OS X?

luke-jr replied:
Ideally, autotools should detect if the platform has MSG_NOSIGNAL and define it if not.

taoeffect replied:
Erm, can't that be done with a simple `#ifdef`...?

phelixbtc replied:
```
#if defined(__WXMSW__) || defined(__clang__)
#define MSG_NOSIGNAL        0
#endif
```

I don't know about a direct define for this. Maybe using 'MAC_OSX' instead of '__ clang __' would be cleaner.

luke-jr replied:
Yes, that's probably safe too -

``` C++
#ifndef MSG_NOSIGNAL
#define MSG_NOSIGNAL 0
#endif
```

crosser replied:
On 09/19/2014 10:23 AM, phelixbtc wrote:

> In src/util.h:
> 
> > @@ -109,12 +109,13 @@ T\* alignup(T\* p)
> >      return u.ptr;
> >  }
> > 
> > +#define MSG_NOSIGNAL        0
> 
> |#if defined(**WXMSW**) || defined(**clang**)
> #define MSG_NOSIGNAL        0
> #endif

This will break build with clang on linux, won't it?

#ifndef MSG_NOSIGNAL

# define MSG_NOSIGNAL 0

#endif

is probably safer, albeit not bullet-proof. On Linux, MSG_NOSIGNAL is a enum
value, but it is accompanied with a #define, making it workable (hopefully).

luke-jr commented on src/serialize.h:
```diff
@@ -926,7 +926,7 @@ class CDataStream
     iterator insert(iterator it, const char& x=char()) { return vch.insert(it, x); }
     void insert(iterator it, size_type n, const char& x) { vch.insert(it, n, x); }
-
+#ifdef __clang__
```
This looks odd/wrong...

taoeffect replied:
What's wrong?

luke-jr replied:
You're essentially removing code based on the compiler... that should never be necessary

taoeffect replied:
If you have some C++ that works everywhere please do share it. I abhor the language and this was good enough to get it to work.

luke-jr commented on src/Makefile.osx:
```diff
@@ -0,0 +1,91 @@
+CXX=g++
```
Why not just use autotools?

taoeffect replied:
There's a PR that uses that I think: https://github.com/namecoin/namecoin/pull/135

The title of this PR explains why I stuck with Make.

luke-jr replied:
I don't see any explanation in the title. Unless you mean "minimal".

taoeffect replied:
Yes, I mean minimal. PR #135 is roughly 60x larger, and nobody seems inclined to audit it.

taoeffect committed:
Bloody MSG_NOSIGNAL defines for linux/mac/win compat.

Thanks for pointing it out @luke-jr!

taoeffect committed:
Fix last breaking change on linux. Now compiles (at least with clang :P)

taoeffect commented:
OK, I actually replaced `__clang__` with `__APPLE__` because ironically enough that broke compilation with clang under linux. Now I've verified it compiles on OS X and Linux using clang. I wasn't able to test gcc on either platform however (OS X doesn't have it, and on linux my VPS kills the process for probably using too much RAM).

phelixbtc commented:
Compiles and runs on Windows. ACK. Linux GCC?

JeremyRand commented:
IMO someone should test on Linux GCC before this is merged.  Maybe someone can ask pmc to look at it?

luke-jr commented:
Just because something compiles does not make it correct. With consensus systems, this is even more important of a point. I advise figuring out the actual issue and fixing it specifically. Or at the very least, leave unaffected platforms unchanged.

JeremyRand commented:
Luke raises a good point.  Is there any reason to believe one way or the other regarding whether this will affect consensus?

phelixbtc commented:
@luke-jr Are you saying this affects any platform other than mac?

pmconrad commented:
https://github.com/okTurtles/namecoin/archive/1f4a78a2ab1e973b96fa6edcff4d2dbc04b10ce0.tar.gz compiles with linux/gcc + starts syncing... I don't see why non-OSX builds should be affected by the patch.

luke-jr commented on src/serialize.h:
```diff
@@ -926,7 +926,7 @@ class CDataStream
     iterator insert(iterator it, const char& x=char()) { return vch.insert(it, x); }
     void insert(iterator it, size_type n, const char& x) { vch.insert(it, n, x); }
-
+#ifdef __APPLE__
```
This change effectively removes the first insert method from the class on non-Apple platforms.

taoeffect replied:
On OS X, the compiler errors because of multiple definitions. I can put the guard around the parameter definition if you'd prefer (as that's the only difference) and delete the duplicate body. Maybe there's a way to define the parameters correctly for all platforms, but again, I have no interest in figuring that out, I just made a few small changes to get it to compile on OS X.

Midar replied:
There is no need for a guard. Just remove the correct duplicate and it should work with any compiler on any OS.

> Maybe there's a way to define the parameters correctly for all platforms, but again, I have no interest in figuring that out, I just made a few small changes to get it to compile on OS X.

Then why bother at all if you only want a quick and dirty hack?!

taoeffect replied:
Alternatively, it might also work to simply show the second definition on all platforms except `__APPLE__`, but so far this seems to work just fine (indicating only one definition is needed... per platform).

taoeffect replied:
> Just remove the correct duplicate and it should work with any compiler on any OS.

Removing one will break compilation on one of the platforms (regardless of which you remove), if I remember correctly.

Feel free to send your own PR on top of this one if you'd like.

> Then why bother at all if you only want a quick and dirty hack?!

...I wanted "minimal changes" to get it to work on OS X. I got them.

Midar commented:
It really saddens me a lot to see how much time people spend to place one hack over another. Things like these `#ifdef __clang__` were not necessary with my PR, it solved everything cleanly. The amount of time you all spent here to just add a few hacks would have been enough to review my whole PR and do it the clean way _and_ get the GUI to work on OS X. With this, there's still no GUI on OS X! It really saddens me to see how resources are wasted for a half-working hack instead of doing it cleanly - and yet people are arguing about what is correct and what not (hint: all that's proposed is not the clean way to do it), even though this is just a collection of hacks and a clean solution has been proposed _and_ done already.

taoeffect commented:
@Midar, I agree with your sentiment with the exception that I believe C++ and autoconf and all that jazz are already hacks by their very nature.

Midar commented:
@taoconf autoconf is not really a hack IMHO. It's the correct way to write portable software: Instead of having checks like "Is this Linux? Is this OSX?", it actually _checks_ if the things it needs are available. That greatly increases portability, as this means it even works on OSes about which the software does not know. The only way where I'd say this is a hack is in that there's nothing integrated with C/C++ that could do these checks, not even a #try_include which would already make a lot of checks unnecessary.

As for C++, yes, you could call it a collection of hacks. For each shortcoming of C or an old C++, a new feature was added. But this is how incremental designs work. Not everything is fully planned ahead and then never extended. In fact, almost nothing is. That's still different from adding even more hacks than necessary if there's a proper solution, though ;). Those hacks in C++ are because either a.) backwards compatibility or b.) there's no better way to do it given the constraints.

taoeffect commented:
> (hint: all that's proposed is not the clean way to do it), even though this is just a collection of hacks and a clean solution has been proposed and done already.

IMO, your PR is an even less clean solution (no offense). It's thousands of additional lines that so far no one has taken the time to review. A C++ node _is_ an ugly, dirty, and potentially dangerous hack. There's no getting around that. This is what I tell everyone when they think it's a bright idea to start new projects in C++ these days. They don't listen. Then this crap happens and we spend all this time and effort on getting crappy, ugly, insecure code to run on multiple platforms. Where's the puke emoticon when you need it?

The "clean" solution would be to write this in a non-hack language.

Midar commented:
The problem here is not C++, but the lack of a build system that allows to properly check for features. @Luke-Jr rightfully complained about those #ifdefs you added, as they are OS checks instead of feature checks. Not even do they limit portability to platforms which are known by the source, this is also a guarantee that things will break in an update of the OS. Your #define for example _will_ break if OS X decides it's a good thing to have it and also adds it.

There are a lot of use cases where C++ is the right tool. Granted, there are better languages for a cryptocurrency, but C++ really isn't the biggest problem here. A way bigger problem is the mix of Boost and Qt, mixing two giant frameworks. Or that there's a lot of `#ifdef __someos__`.

Those additional lines suddenly get down to a _really_ low number if you just compare those files that are taken from autotools with the upstream ones. It gets down to a few hundred lines. Really now, if you don't trust autotools from upstream, you can't trust your system at all. Heck, your compiler uses them, so why even trust your compiler? Anyway, the amount of time spent on this PR is already magnitudes larger than the amount of time that would have been required to review those parts of my PR that are not just copied from elsewhere (and those you can easily diff), while the gain is significantly smaller.

phelixbtc commented:
@Midar Namecoin will be rebased on current Bitcoin / Libcoin. There is a consensus that we will not add a build system at this point. Please take further discussions about this to the forum or make a PR for contrib/

I appreciate all of you staying polite and it may be a bit late as we hopefully are about done here, but still: Everybody please stay on topic. Off topic posts will be deleted from now on. 

@taoeffect This is just what we needed. Thank you!

indolering commented:
@luke-jr @Midar We **really** appreciate both of your efforts on this patch, we don't want to impact non-OS X systems!  Remember that we are just trying to make the minimum changes required to get this building on OS X.

taoeffect commented:
> @luke-jr @Midar We really appreciate both of your efforts on this patch, we don't want to impact non-OS X systems! Remember that we are just trying to make the minimum changes required to get this building on OS X.

Ditto. :+1:

Midar commented:
@indolering Sometimes, minimal effort equals to the most dirty solution. While my patch does touch other OSes, it does not break any. In fact, it even increases portability for other OSes.
